dsc_lib/src/progress.rs

```diff
@@ -0,0 +1,192 @@
+    /// Create a `ProgressBar` object to update progress
+    ///
+    /// # Arguments
+    ///
+    /// * `item_count` - Total number of steps to complete.  Use '1' if unknown and increment when complete.
+    /// * `format` - The `ProgressFormat` to render.
+    ///
+    /// # Returns
+    ///
+    /// A `ProgressBar` oject to update progress
+    ///
+    /// # Errors
+    ///
+    /// Fails if progress style for console rendering can't be set.
+    ///
+    pub fn new(item_count: u64, format: ProgressFormat) -> Result<ProgressBar, DscError> {
+        let bar = warn_span!("");
+        if format == ProgressFormat::Default {
+            bar.pb_set_style(&ProgressStyle::with_template(
+                "{spinner:.green} [{elapsed_precise:.cyan}] [{bar:40.cyan/blue}] {pos:>7}/{len:7} {msg:.yellow}"
+            )?);
+            bar.pb_set_length(item_count);
+            let _guard = bar.enter();
+        }
+
+        Ok(ProgressBar {
+            progress_value: Progress::new(),
+            console_bar: bar,
+            item_count,
+            item_position: 0,
+            format
+        })
+    }
+
+    pub fn increment(&mut self, delta: u64) {
+        if self.format == ProgressFormat::None {
+            return;
+        }
+
+        self.item_position += delta;
+
+        if self.item_count  > 0 {
+            self.progress_value.percent_complete = if self.item_position >= self.item_count {
+                100
+            } else {
+                u8::try_from((self.item_position * 100) / self.item_count).unwrap_or(100)
+            };
+        }
+
+        if self.format == ProgressFormat::Json {
+            self.write_json();
+        } else {
+            self.console_bar.pb_inc(delta);
+        }
+    }
+
+    pub fn set_resource(&mut self, name: &str, resource_type: &str, result: Option<&Value>) {
+        self.progress_value.resource_name = Some(name.to_string());
+        self.progress_value.resource_type = Some(resource_type.to_string());
+        self.progress_value.result = result.cloned();
+    }
+
+    pub fn set_activity(&mut self, activity: &str) {
+        match self.format {
+            ProgressFormat::Json => {
+                self.progress_value.activity = Some(activity.to_string());
+                self.write_json();
+            },
+            ProgressFormat::Default => {
+                self.console_bar.pb_set_message(activity);
+            },
+            ProgressFormat::None => {}
+        }
+    }
+
+    pub fn set_length(&mut self, len: u64) {
+        match self.format {
+            ProgressFormat::Json => {
+                self.item_count = len;
+                if self.item_count  > 0 {
+                    self.progress_value.percent_complete = if self.item_position >= self.item_count {
+                        100
+                    } else {
+                        u8::try_from((self.item_position * 100) / self.item_count).unwrap_or(100)
+                    };
+                }
```

Comment:
this might be intentional but just noting that `increment()` and `set_position()` call `write_json()` after updating `self.progress_value.percent_complete`

Reply:
Only change in activity and increment of progress results in writing progress.  I'll add comments to the public functions, surprised clippy didn't complain